### PR TITLE
Don't suggest incorrect syntax

### LINKED ISSUE #51634: misleading help message for ambiguous numeric type in for loop over a range
The following code ([playground](https://play.rust-lang.org/?gist=77a4cfb33621b739973cb3d15d08069b&version=nightly&mode=debug)):

    for i in 0..100 {
        println!("{}", i.pow(2));
    }

fails to compile. The error message correctly points to the problem but the help message's suggestion is a syntax error:

    error[E0689]: can't call method `pow` on ambiguous numeric type `{integer}`
     --> src/main.rs:3:26
      |
    3 |         println!("{}", i.pow(2));
      |                          ^^^
    help: you must specify a type for this binding, like `i32`
      |
    2 |     for i: i32 in 0..100 {
      |         ^^^^^^

Following the suggestion leads to another compile error (missing 'in' in for loop). Usually the correct solution in this case is to add the type to the range (`0i32..100`).

### PR DESCRIPTION
**Pull request by estebank**
Fix #51634.

### CODE REVIEW

**Comment by estebank** [comment]
Can you provide a repro case? I'll see if I have any way of differentiate them (other than checking the span context).

**Comment by oli-obk** [comment]
somewhat xD I found other stuff:

```rust
    let x: _ = 42;
    x.pow(2);
```

```
help: you must specify a type for this binding, like `i32`
  |
8 |     let x: i32: _ = 42;
  |         ^^^^^^
```

and

```rust
macro_rules! foo {
    ($ident:ident) => { let $ident = 42; }
}

fn main() {
    foo!(bar);
    bar.pow(2);
}
```

suggests to modify a macro (which might be not from this crate)

```
help: you must specify a type for this binding, like `i32`
  |
2 |     ($ident:ident) => { let $ident: i32 = 42; }
  |                             ^^^^^^^^^^^ 
```

**Comment by estebank** [comment]
@oli-obk as is, on external macros, the suggestion uses the compiled macro output:

```
error[E0689]: can't call method `pow` on ambiguous numeric type `{integer}`
  --> $DIR/method-on-ambiguous-numeric-type.rs:40:9
   |
LL |     bar.pow(2);
   |         ^^^
help: you must specify a type for this binding, like `i32`
   |
LL | ( $ ident : ident ) => { let $ ident: i32 = 42 ; }
   |                              ^^^^^^^^^^^^
```

**Comment by estebank** [comment]
Changed to account for both of the presented cases.

File: src/librustc_typeck/check/method/suggest.rs [hunk 1]
```diff
@@ -271,13 +277,18 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
                                             "you must specify a type for this binding, like `{}`",
                                             concrete_type,
                                         );
-                                        match parent_node {
-                                            hir_map::NodeLocal(hir::Local {
+
+                                        match (is_real_filename, parent_node) {
```
Comment (oli-obk) [inline src/librustc_typeck/check/method/suggest.rs]: You match on filename above toget this bool. Can you just match on it directly here?

**Review by oli-obk — CHANGES_REQUESTED** [review]

File: src/librustc_typeck/check/method/suggest.rs (hunk 1 above)
Comment (estebank) [reply src/librustc_typeck/check/method/suggest.rs]: Done.

**Review by oli-obk — APPROVED** [review]

**Pull request merged**